Page Orientation

PerlTester opened the pull request:
This commit adds the necessary code to allow Portrait and Landscape page
changes within a PDF document.

Implemented a new PDF Object parameter --page-orientation with a value of P for portrait and L for landscape.

Given 3 HTML files where you want First Landscape, Second Portrait and Third Landscape invoke as follows:

wkhtmltopdf.exe part_1.htm --page-orientation L part_2.htm --page-orientation P part_3.htm --page-orientation L Output.pdf

ashkulz commented:
This seems to be crashing for me on Windows, will try it again on Linux and see what the issue is.

PerlTester commented:
I've just rebuilt it again on debian with the cross compiler for windows and it works on Win2008R2 and Windows 10 so not sure why its crashing for you.

Pat.

Zipped Exe attached.

[wkhtmltox-0.12.4-dev-orientation-4fa8338_mingw-w64-cross-win64.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/275450/wkhtmltox-0.12.4-dev-orientation-4fa8338_mingw-w64-cross-win64.zip)

PerlTester commented:
Any Luck?

Pat

On Sun, May 15, 2016 at 8:30 PM, Ashish Kulkarni <notifications@github.com>
wrote:

> This seems to be crashing for me on Windows, will try it again on Linux
> and see what the issue is.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/wkhtmltopdf/wkhtmltopdf/pull/2946#issuecomment-219302266>
>

chri3g91 commented:
Hi PerlTester, 
I have tried your update (0.12.4) with the print orientation in an ASP.NET web application and am having an issue. I have successfully implemented wkhtmltopdf version 0.12.3.1 to a create multi-page PDF report, but would like the option for page orientation on a per-page selection. 
I can create a single page, no problem, but when I have multiple html pages, the PDF is being created and looks normal in a "wktemp" file, but is not copying correctly to the output file that is downloaded in the browser. 
You can clearly see an error in the size of the PDFs. Any idea why??

![screenshot_1](https://cloud.githubusercontent.com/assets/7135767/18991413/b0c34a46-8718-11e6-84d9-fb8bbdb9df5d.png)
![screenshot_2](https://cloud.githubusercontent.com/assets/7135767/18991414/b0c7f618-8718-11e6-8f8d-ab62df8abb9e.png)

PerlTester commented:
Hi Chris,

Sorry I am not following :-(

So have you complied and produced a windows version of wkhtmltopdf based on
the edits I provided for the page orientation options?

Or are you just using the windows version I posted ?

wkhtmltox-0.12.4-dev-orientation-4fa8338_mingw-w64-cross-win64.zip
<https://github.com/wkhtmltopdf/wkhtmltopdf/files/275450/wkhtmltox-0.12.4-dev-orientation-4fa8338_mingw-w64-cross-win64.zip>

Either way can you try in command line mode to add multiple html pages
together by prefixing the option for either portrait or landscape along the
lines of

wkhtmltopdf.exe part_1.htm --page-orientation L part_2.htm
--page-orientation P part_3.htm --page-orientation L Output.pdf

where you have 3 html files ready to convert to pdf - namely part_1.htm,
part_2.htm and part_3.htm

in the above example you should produce a combined 3 page pdf with the
first page landscape, the second portrait and the third landscape and the
resulting pdf will be called Output.pdf

if you get the above working then you should just shell to this command
line from ASP.NET - remember prior to shelling you need to have split the
html files so you can tag them with the correct orientation L or P as
required. You then slurp the resulting file and return to the browser.

Not sure if that helps or is there something I am missing that you are
doing/not doing?

Cheers

P.










On Fri, Sep 30, 2016 at 2:18 PM, chri3g91 <notifications@github.com> wrote:

> Hi PerlTester,
> I have tried your update (0.12.4) with the print orientation in an ASP.NET
> web application and am having an issue. I have successfully implemented
> wkhtmltopdf version 0.12.3.1 to a create multi-page PDF report, but would
> like the option for page orientation on a per-page selection.
> I can create a single page, no problem, but when I have multiple html
> pages, the PDF is being created and looks normal in a "wktemp" file, but is
> not copying correctly to the output file that is downloaded in the browser.
> You can clearly see an error in the size of the PDFs. Any idea why??
>
> [image: screenshot_1]
> <https://cloud.githubusercontent.com/assets/7135767/18991413/b0c34a46-8718-11e6-84d9-fb8bbdb9df5d.png>
> [image: screenshot_2]
> <https://cloud.githubusercontent.com/assets/7135767/18991414/b0c7f618-8718-11e6-8f8d-ab62df8abb9e.png>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wkhtmltopdf/wkhtmltopdf/pull/2946#issuecomment-250730329>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACk-NbAT2fWvv3OjPpJNubcSbYiSDC16ks5qvP4ggaJpZM4Id0KS>
> .
>

chri3g91 commented:
Hi again PearlTester, 

Very sorry that I never got back on here to give you a response. I moved on to other tasks, but now it is time to re-address this issue.  For clarity, I am using the windows version that you posted. As I stated before, it works when printing single pages, but the correct output file in the browser. As I showed in the images form my previous message, the library creates this temp file that is correctly displaying the report with the declared page orientation. but the file "print" is not. 

I tested the arguments  inside command line as well and I am getting the correct outputs. The image shows the code snippet that I am using to write the file back to the user, where it can be downloaded. This works fine for multiple page reports using wkhtmltopdf version 0.12.3. but not your uploaded version. So again I ask, do you have any ideas and/or can you clarify what you mean by 'slurping the result'?

![read output snip](https://cloud.githubusercontent.com/assets/7135767/20030732/1e10cb14-a36b-11e6-853e-b3554ecab6da.png)

Thanks again!

PerlTester commented:
Hi Chris,

Sorry again but I have only used this within my perl applications where I
generate the commandline and execute it via a system call. When it
completes I pick up the resultant output file and serve it back to the
browser (I'm using the Perl Catalyst framework). I am not familiar with
asp.net i'm afraid. This works for me. Are you calling pdf creation via api
calls or via shelling to a command line?

The only changes I made to the source code was to chcek for the P and L
switches and change the print orientation other than that Im afraid Im no
help to you.

heres how I use it in my perl code

>>>>>> cut from code >>>>>

#### Split into mutiple files based on <!-- NEW PAGE --> found in
preparation for Landscape fix to wkhtml2pdf
my @part_files=();
my $partfile_list=qq{};
my $css_links=qq{};
my $webkit_commandline_options=qq{};
my $portrait_landscape_options=qq{};

my $counter=0;
my $output_part_file_0 =
$c->config->{home}.qq{/root/htmldoc/temp/}.$module_code.qq{_}.$first_intake.qq{_part_}.$counter.qq{\.html};

my $INFILE = new FileHandle;
my $OUTFILE = new FileHandle;

use Fcntl;
open($INFILE, "<".$input_file) or die "Could not open file $!";
#$c->log->debug(qq{***Files : Opening Input File}.$input_file);

open($OUTFILE,">".$output_part_file_0) || die "couldn't open
$output_part_file_0: $!";
#$c->log->debug(qq{***Files : Opening }.$output_part_file);

while(<$INFILE>) {

  # if line contains with '<!-- Landscape -->' or '<!-- Portrait -->',
split the file
 if (m/<!-- Landscape -->/ || m/<!-- Portrait -->/) {

    close($OUTFILE);
    $counter++;
    my $output_part_file =
$c->config->{home}.qq{/root/htmldoc/temp/}.$module_code.qq{_}.$first_intake.qq{_part_}.$counter.qq{\.html};

    $portrait_landscape_options=qq{};

    if (m/<!-- Landscape -->/) {
    $portrait_landscape_options=qq{--page-orientation L};
    }
    if (m/<!-- Portrait -->/) {
    $portrait_landscape_options=qq{--page-orientation P};
    }

    push(@part_files, qq{"$output_part_file" }.$portrait_landscape_options);

    open($OUTFILE,">".$output_part_file) || die "couldn't open
$output_part_file: $!";
    #$c->log->debug(qq{***Files : Opening }.$output_part_file);
  #$c->log->debug(qq{***Printing: CSS }.$css_links);
    print $OUTFILE $css_links;

  } else {

  if (/<!-- --(.*?)-->/) {  # is there a comment on this line with a
--configword
    $c->log->debug(qq{***Webkit: Comment Match --}.$1.qq{\n});
    $webkit_commandline_options.=qq{ --}.$1.qq{ };
    }

    if (m/<link rel="stylesheet" href=/) {
    $css_links.=$_.qq{\n};
    }
  #$c->log->debug(qq{***Printing: \$_ }.$_);
    print $OUTFILE $_;
  }

}

close($INFILE);
close($OUTFILE);

foreach(@part_files){
$partfile_list .= $_.qq{ }.$webkit_commandline_options.qq{ };
}

$c->log->debug(qq{***WebKit: }.qq{"$filepath" $title_file $template_toc
$template_htmldoc_settings "$output_part_file_0"
$webkit_commandline_options $partfile_list --load-error-handling ignore
"$output_file"});

system qq{"$filepath" $title_file $template_toc $template_htmldoc_settings
"$output_part_file_0" $webkit_commandline_options $partfile_list
--load-error-handling ignore "$output_file"};

}

# Suck the PDF contents back in so we can serve it up via the controller
# we use do{}; and local undef $/; ($/ is the newline separator function)
so as to
# not kill the global $/ operator

open(FILE,$output_file);
my $pdf_file = do { binmode FILE ; local $/; <FILE> };
close (FILE);

#$c->log->debug(qq{***PDFDoc: $pdf_file});

 $c->response->content_type('application/pdf; charset=UTF-8');
 $c->response->header('Content-Disposition', qq[attachment;
filename="$module_code\_$first_intake$template_title\.pdf"]);
 $c->response->body($pdf_file);






On Sat, Nov 5, 2016 at 3:23 PM, chri3g91 <notifications@github.com> wrote:

> Hi again PearlTester,
>
> Very sorry that I never got back on here to give you a response. I moved
> on to other tasks, but now it is time to re-address this issue. For
> clarity, I am using the windows version that you posted. As I stated
> before, it works when printing single pages, but the correct output file in
> the browser. As I showed in the images form my previous message, the
> library creates this temp file that is correctly displaying the report with
> the declared page orientation. but the file "print" is not.
>
> I tested the arguments inside command line as well and I am getting the
> correct outputs. The image shows the code snippet that I am using to write
> the file back to the user, where it can be downloaded. This works fine for
> multiple page reports using wkhtmltopdf version 0.12.3. but not your
> uploaded version. So again I ask, do you have any ideas and/or can you
> clarify what you mean by 'slurping the result'?
>
> [image: read output snip]
> <https://cloud.githubusercontent.com/assets/7135767/20030732/1e10cb14-a36b-11e6-853e-b3554ecab6da.png>
>
> Thanks again!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wkhtmltopdf/wkhtmltopdf/pull/2946#issuecomment-258614664>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACk-NTLC75zU4cfZv2emW1cRz5_XkA53ks5q7JFEgaJpZM4Id0KS>
> .
>

elbenfreund commented:
Hi there. Is there any general news on this, besides this individual issue?

PerlTester commented:
Nothing from my side, its as I left it with the maintainer who said it was
crashing for him, I was hoping it would make it into the build but nothing
more I can do.

Cheers

P.


On Tue, Feb 28, 2017 at 2:28 PM, Eric Goller <notifications@github.com>
wrote:

> Hi there. Is there any general news on this, besides this individual issue?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wkhtmltopdf/wkhtmltopdf/pull/2946#issuecomment-283038456>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACk-NadN44fJ-2-v8_JVzWj50qLLJ_L6ks5rhCEEgaJpZM4Id0KS>
> .
>

andrehsmendes commented:
@ashkulz 

Any ups from your side?

thx for the great repo.

andrehsmendes commented:
@PerlTester 

Can you add support for options like so:

--page-orientation <page> <orientation>

Greatly appreciated.

PerlTester commented:
Hi Andre,

Sorry but I don't have the time at the moment to add support for what you suggest, I added the original pull request over a year ago and still no joy, pity. What I have works well for my use case at the moment.

P.

andrehsmendes commented:
The executable file with the new parameters is working fine for me on windows 10. no issues or crashes so far.
@ashkulz 

why not to merge this yet?

andrehsmendes commented:
@PerlTester 

Hey my friend, can you compile a version of your changes for ubuntu server? Greatly appreciate.

Andre

PerlTester commented:
No Problem,

Here you go. Unzip it first to a /home/install/ directory and then install it will the following

echo Installing/Updating WKHTMLTOPDF from local package...
apt-get -y install openssl build-essential xorg libssl-dev
tar -xvf /home/install/wkhtmltox-0.12.4-dev-orientation-4fa8338_linux-generic-amd64.tar.xz
cp -r /home/install/wkhtmltox/. /usr/local/
rm -r -f /home/install/wkhtmltox/
echo Fixing dependencies...
apt-get -y install -f

Let us know how you get on

P.

[wkhtmltox-0.12.4-dev-orientation-4fa8338_linux-generic-amd64.tar.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/1534698/wkhtmltox-0.12.4-dev-orientation-4fa8338_linux-generic-amd64.tar.zip)

andrehsmendes commented:
@PerlTester 

The installation went fine. Thank you for that.

Now im getting this error.

`[2017-12-06T18:31:17.315] [ERROR] wkhtml - [ '/home/andre/proyekto/tmp/htmlpd                  f/0_1512585077313.html',
  '/home/andre/proyekto/tmp/htmlpdf/1_1512585077314.html',
  '/home/andre/proyekto/tmp/htmlpdf/2_1512585077314.html',
  '/home/andre/proyekto/tmp/htmlpdf/3_1512585077314.html',
  '/home/andre/proyekto/tmp/htmlpdf/4_1512585077314.html',
  '/home/andre/proyekto/tmp/htmlpdf/5_1512585077314.html',
  '/home/andre/proyekto/tmp/htmlpdf/6_1512585077314.html',
  '/home/andre/proyekto/tmp/htmlpdf/7_1512585077315.html',
  '/home/andre/proyekto/tmp/htmlpdf/8_1512585077315.html' ]
[2017-12-06T18:31:17.576] [ERROR] wkhtml - stderr: QPainter::begin(): Returne                  d false

[2017-12-06T18:31:17.576] [ERROR] wkhtml - stderr: Error: Unable to write to                   destination

[2017-12-06T18:31:17.576] [ERROR] wkhtml - stderr: Exit with code 1, due to u                  nknown error.

[2017-12-06T18:31:17.579] [ERROR] wkhtml - Error Code: 1
`
Do you think it is related to file permissions?

Andre

PerlTester commented:
looks that way to be honest

On Wed, Dec 6, 2017 at 7:37 PM, Andre Mendes <notifications@github.com>
wrote:

> @PerlTester <https://github.com/perltester>
>
> The installation went fine. Thank you for that.
>
> Now im getting this error.
>
> `[2017-12-06T18:31:17.315] [ERROR] wkhtml - [ '/home/andre/proyekto/tmp/htmlpd
> f/0_1512585077313.html',
> '/home/andre/proyekto/tmp/htmlpdf/1_1512585077314.html',
> '/home/andre/proyekto/tmp/htmlpdf/2_1512585077314.html',
> '/home/andre/proyekto/tmp/htmlpdf/3_1512585077314.html',
> '/home/andre/proyekto/tmp/htmlpdf/4_1512585077314.html',
> '/home/andre/proyekto/tmp/htmlpdf/5_1512585077314.html',
> '/home/andre/proyekto/tmp/htmlpdf/6_1512585077314.html',
> '/home/andre/proyekto/tmp/htmlpdf/7_1512585077315.html',
> '/home/andre/proyekto/tmp/htmlpdf/8_1512585077315.html' ]
> [2017-12-06T18:31:17.576] [ERROR] wkhtml - stderr: QPainter::begin():
> Returne d false
>
> [2017-12-06T18:31:17.576] [ERROR] wkhtml - stderr: Error: Unable to write
> to destination
>
> [2017-12-06T18:31:17.576] [ERROR] wkhtml - stderr: Exit with code 1, due
> to u nknown error.
>
> [2017-12-06T18:31:17.579] [ERROR] wkhtml - Error Code: 1
> `
> Do you think it is related to file permissions?
>
> Andre
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wkhtmltopdf/wkhtmltopdf/pull/2946#issuecomment-349733718>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACk-NdxpvRrce4A-r0bHVlCzT9yxh9ucks5s9t7bgaJpZM4Id0KS>
> .
>

andrehsmendes commented:
@PerlTester 

I've been searching and looks like its a problem with STDOUT.

On my node application i'm returning it as a stream from STDOUT instead of generating a file.

Piece of the code:

`      const makePdf = spawn('wkhtmltopdf', args);
        // Stream the results
        var stream = makePdf.stdout;

        makePdf.stderr.on('data', (data) => {
            logger.error(`stderr: ${data.toString()}`);
        });

        // On exit pipe stream to response
        makePdf.on('exit', (code) => {
            if (code !== 0) {
                console.log('Error', code)
                logger.error('Error Code: '+ code);
            } else {
                // Remove temp files
               
                for (let index = 0; index < htmlLocation.length; index++) {
                    const element = htmlLocation[index];
                    fs.unlink(element, err => {
                        if (err) throw err;
                    });
                }
               
                res.setHeader('Content-disposition', `inline; filename="Production Document"`)
                res.setHeader('Content-type', 'application/pdf')
                stream.pipe(res)
            }
        });`

PerlTester commented:
ok but is it just for my build or are you getting the same issue for the standard build?

andrehsmendes commented:
@PerlTester 
I got ir working now.

The problem was the parameters that i was adding to add space from the headers.

It is very weird because locally it works just fine with the parameters. I believe that there is something to do with how the web-kit emulates on the server.

Before:
```
let args = [
            "--quiet",
            "--page-size", "letter",
            "--header-html", htmlLocation[0],
            "--header-spacing", "5",
            "--margin-left", "5",
            "--margin-right", "5",
            "--margin-top", "68",
            "--checkbox-checked-svg", checkedsvg,
            "--checkbox-svg", checksvg,
            "--zoom", "2"
        ];
```
Removed:
```
 "--header-spacing", "5",
 "--margin-top", "68",
```

PohlodSlavik commented:
Hi guys,

There are more than 2 years passed from pull request creation.
Is there any chance that this pull request will be merged to the main branch?
It would be really nice to have this feature included.

@ashkulz , what is actually a problem with merging this to main branch?

andrehsmendes commented:
yes. i don't know why it was not merged with the main branch yet.

PerlTester commented:
As this code was written over 3 years ago and wkhtmltopdf has moved on I am closing this pull request. I will reopen a new pull request with code written for the latest branch.

Perltester
